kie-server-parent/kie-server-services/kie-server-services-jbpm-ui/src/main/java/org/kie/server/services/jbpm/ui/FormRendererBase.java

```diff
@@ -0,0 +1,219 @@
+        String resolvedContainerId = registry.getContainerId(containerId, new ByTaskIdContainerLocator(taskId));
+
+        Task task = userTaskService.execute(resolvedContainerId, new GetUserTaskCommand(registry.getIdentityProvider().getName(), taskId));
+        if (task == null) {
+            throw new TaskNotFoundException("No task with id " + taskId + " found");
+        }
+        String name = getTaskFormName(task);
+
+        String formStructure = formManagerService.getFormByKey(resolvedContainerId, name);
+        if (formStructure == null) {
+            throw new IllegalArgumentException("Form for task " + taskId + " not found with supported suffix " + SUPPORTED_FORM_SUFFIX);
+        }
+
+        FormInstance form = formReader.readFromString(formStructure);
+        form.setNestedFormsLookup(formId -> {
+            Map<String, String> forms = indexedForms.get(resolvedContainerId);
+            String formStructureNested = forms.get(formId);
+            return formReader.readFromString(formStructureNested);
+        });
+        
+        Map<String, Object> inputData = userTaskService.getTaskInputContentByTaskId(resolvedContainerId, taskId);        
+        Map<String, Object> outputData = userTaskService.getTaskOutputContentByTaskId(resolvedContainerId, taskId);        
+        
+        FormRenderer formRenderer = renderers.get(renderer);
+        if (formRenderer == null) {
+            throw new IllegalArgumentException("Form renderer with name " + renderer + " not found");
+        }
+        String output = formRenderer.renderTask(resolvedContainerId, task, form, inputData, outputData);
+        
+        return output;
+    }
+    
+    public String getCaseRenderedForm(String renderer, String containerId, String caseDefinitionId) {
+        if (caseRuntimeDataService == null) {
+            throw new RuntimeException("Case capability in jBPM UI are not enabled");
+        }
+        String resolvedContainerId = registry.getContainerId(containerId, ContainerLocatorProvider.get().getLocator());
+
+        CaseDefinition caseDefinition = caseRuntimeDataService.getCase(resolvedContainerId, caseDefinitionId);
+        if (caseDefinition == null) {
+            throw new CaseDefinitionNotFoundException("Case definition " + resolvedContainerId + " : " + caseDefinitionId + " not found");
+        }
+        
+        String formStructure = formManagerService.getFormByKey(resolvedContainerId, caseDefinitionId + SUPPORTED_FORM_SUFFIX);
+        
+        
+        FormInstance form = formReader.readFromString(formStructure);
+        if (form != null) {
+            form.setNestedFormsLookup(formId -> {
+                Map<String, String> forms = indexedForms.get(resolvedContainerId);
+                String formStructureNested = forms.get(formId);
+                return formReader.readFromString(formStructureNested);
+            });
+        }
```

Comment:
it can't be easily externalised as it must refer to the context of the calling method like container id etc. And yes, case can be started without data but worth to have case role assignment set